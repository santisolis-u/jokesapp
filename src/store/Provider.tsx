import React, {useEffect, useState} from 'react';
import {Joke} from '../hooks/useJokes';
import {FavJokesContext} from './store';
import Realm from 'realm';
import {JokeSchema} from '../realm/schema';
interface Props {
  children: JSX.Element;
}

export const FavsProvider: React.FC<Props> = ({children}) => {
  const [favJokes, setFavJokes] = useState<Joke[]>([]);
  const [realm, setRealm] = useState<Realm>();

  useEffect(() => {
    const openRealm = async () => {
      const r = await Realm.open({schema: [JokeSchema]});
      if (r != null && !r.isClosed) {
        setRealm(r);
        const result = r.objects<Joke>('Joke').map(value => {
          return value.toJSON() as Joke;
        });
        setFavJokes(result);
      }
    };
    openRealm();
    return () => {
      if (!!realm && !realm.isClosed) {
        realm.close();
      }
    };
  }, []);

  const addFavJoke = (joke: Joke) => {
    if (!favJokes.includes(joke)) {
      if (!!realm && !realm.isClosed) {
        realm.write(() => {
          realm.create('Joke', joke);
        });
      }
      setFavJokes(favJokes => [...favJokes, joke]);
    }
  };
  const removeFavJoke = (id: string) => {
    if (!!realm && !realm.isClosed) {
      realm.write(() => {
        const joke = realm.objectForPrimaryKey<Joke>('Joke', id);
        realm.delete(joke);
      });
    }
    setFavJokes(favJokes => favJokes.filter(joke => joke.id !== id));
  };
  return (
    <FavJokesContext.Provider value={{favJokes, addFavJoke, removeFavJoke}}>
      {children}
    </FavJokesContext.Provider>
  );
};
