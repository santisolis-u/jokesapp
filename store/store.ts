import {Joke} from '../src/hooks/useJokes';
import React, {FC, ReactChildren, useState} from 'react';
interface FavJokesContextType {
  favJokes: Joke[];
  addFavJoke?: (joke: Joke) => void;
  removeFavJoke?: (id: string) => void;
}

const defaultState = {
  favJokes: [],
};

export const FavJokesContext =
  React.createContext<FavJokesContextType>(defaultState);
