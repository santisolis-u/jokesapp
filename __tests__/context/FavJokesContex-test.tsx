import React from 'react';
import {Pressable, Text} from 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {fireEvent, render} from '@testing-library/react-native';
import {FavsProvider} from '../../src/store/Provider';
import {FavJokesContext} from '../../src/store/store';

const mockedJoke = {
  id: '1',
  joke: 'test',
  permalink: '1',
};

describe('Fav Context', () => {
  test('Renders correctly', () => {
    const component = renderer.create(
      <FavsProvider>
        <FavJokesContext.Consumer>
          {value => (
            <>
              <Text>Loading:{value.favJokes.length}</Text>
            </>
          )}
        </FavJokesContext.Consumer>
      </FavsProvider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  test('Default state on savedJokes', () => {
    const {getByText} = render(
      <FavsProvider>
        <FavJokesContext.Consumer>
          {value => <Text>savedJokes: {value.favJokes.length}</Text>}
        </FavJokesContext.Consumer>
      </FavsProvider>,
    );
    expect(getByText('savedJokes: 0')).toBeTruthy();
  });
  test('function addFavJokes adds a joke to favorites', () => {
    const {getByTestId, getByText} = render(
      <FavsProvider>
        <FavJokesContext.Consumer>
          {value => (
            <>
              <Text>savedJokes: {value.favJokes.length}</Text>
              <Pressable
                onPress={value.addFavJoke?.bind(this, mockedJoke)}
                testID={'press-save'}>
                <Text>Press Save</Text>
              </Pressable>
            </>
          )}
        </FavJokesContext.Consumer>
      </FavsProvider>,
    );
    expect(getByText('savedJokes: 0')).toBeTruthy();
    fireEvent(getByTestId('press-save'), 'press');
    expect(getByText('savedJokes: 1')).toBeTruthy();
  });

  test('function removeFavJoke removes a joke to favorites', () => {
    const {getByText, getByTestId} = render(
      <FavsProvider>
        <FavJokesContext.Consumer>
          {value => (
            <>
              <Text>savedJokes: {value.favJokes.length}</Text>
              <Pressable
                onPress={value.addFavJoke?.bind(this, mockedJoke)}
                testID={'press-save'}>
                <Text>Press Save</Text>
              </Pressable>
              <Pressable
                onPress={value.removeFavJoke?.bind(this, '1')}
                testID={'press-delete'}>
                <Text>Press Delete</Text>
              </Pressable>
            </>
          )}
        </FavJokesContext.Consumer>
      </FavsProvider>,
    );
    expect(getByText('savedJokes: 0')).toBeTruthy();
    fireEvent(getByTestId('press-save'), 'press');
    expect(getByText('savedJokes: 1')).toBeTruthy();
    fireEvent(getByTestId('press-delete'), 'press');
    expect(getByText('savedJokes: 0')).toBeTruthy();
  });
  test('function addFavJokes doesnt add a joke to favorites if it has been already added', () => {
    const {getByTestId, getByText} = render(
      <FavsProvider>
        <FavJokesContext.Consumer>
          {value => (
            <>
              <Text>savedJokes: {value.favJokes.length}</Text>
              <Pressable
                onPress={value.addFavJoke?.bind(this, mockedJoke)}
                testID={'press-save'}>
                <Text>Press Save</Text>
              </Pressable>
            </>
          )}
        </FavJokesContext.Consumer>
      </FavsProvider>,
    );
    expect(getByText('savedJokes: 0')).toBeTruthy();
    fireEvent(getByTestId('press-save'), 'press');
    fireEvent(getByTestId('press-save'), 'press');
    expect(getByText('savedJokes: 1')).toBeTruthy();
  });
});
