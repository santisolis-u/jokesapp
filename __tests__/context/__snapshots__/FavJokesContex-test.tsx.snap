// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fav Context Renders correctly 1`] = `
<Text>
  Loading:
  0
</Text>
`;
